### ISSUE TITLE
containers: support derivation

### ISSUE THREAD

**Issue by cgwalters** [ostreedev/ostree-rs-ext#12]
We may in the future support the equivalent of a `Dockerfile` that does
```
FROM quay.io/exampleos/exampleos:latest
RUN curl ...
ADD ...
RUN ostree container-finalize-bootable
```

What this will do is basically an ostree-tar version of `ostree commit --tree=dir=/ --selinux-policy-from-base` that will do e.g. selinux labeling, hardlink deduplication and computing the sha256 digest.

In theory, we don't require this "blessing" of `ostree container-finalize-bootable` - we could do it client side too.  But I think it's better for the client side to be "dumb" by default - it makes it more like an image system.

## Use/integration with rpm-ostree and higher level projects

This project is independent of rpm-ostree and other tools.  However we need to support them; specifically e.g.:

```
FROM quay.io/coreos/fedora:stable
RUN rpm-ostree install cowsay
RUN ostree container-finalize-bootable
```

Following the above.  And so our library/model should also support what rpm-ostree natively wants to do in unpacking rpms into ostree commits for example.  

Ultimately, this scope creeps basically into a new ostree repo mode `container-tar` or so?

**Comment by cgwalters** [comment]
If we want to support `Dockerfile` then it basically means we need to understand container layer deltas.

**Comment by cgwalters** [comment]
Thinking through this more, it seems to me that derivation support does scope creep into mapping container image layers into ostree commits, and handling GC.  Basically to gain the real benefit, we need to avoid re-pulling layers we already have.

This isn't hard at all, and similar code exists in containers/storage's ostree backend.  But it is definitely a scope increase.

**Comment by cgwalters** [comment]
OK to try this out using e.g. FCOS stable:

First, switch to a build I uploaded that has https://github.com/ostreedev/ostree-rs-ext/pull/99 plus an updated rpm-ostree, plus https://github.com/cgwalters/container-image-proxy stuck in the image too.

These instructions have been tested on current FCOS stable `34.20210904.3.0`.  But note that the syntax for rebasing to containers has changed in the newer rpm-ostree v2021.11 (with newer ostree-rs-ext) which is included in this test container I've uploaded:

```
# systemctl mask --now zincati
# rpm-ostree rebase --experimental docker://quay.io/cgwalters/fcos-dev:latest
...
# systemctl reboot
```

Now from there, try creating a derived image!  You *could* do it on the same machine, but I think it's much clearer to use a separate build system, which could be the quay.io buildsystem or e.g. an OpenShift cluster or whatever.

There's an example here https://github.com/cgwalters/fcos-derivation-example

Right now you can add files - notably things like systemd units.  What doesn't work yet is `yum install`...there's some nontrivial work to do on the rpm-ostree side for that.

Anyways, once you have your image uploaded somewhere, try rebasing to it:

```
# rpm-ostree rebase --experimental ostree-unverified-image:docker://quay.io/cgwalters/fcos-dev:derived
```

From there you can apply via e.g. `systemctl reboot` or alternatively try `rpm-ostree ex apply-live`!

**Comment by mkenigs** [comment]
Would a `Dockerfile` be used or some other format? Just thinking about making MCO/MCBS more declarative and wondering how this would fit in.

**Comment by cgwalters** [comment]
See https://github.com/coreos/enhancements/pull/7/files#diff-bc7b9e656253160761d8ef791b3e206a1264197112e45916edfa9344e97359e7R37

**Comment by cgwalters** [comment]
But to clarify, this proposal does not require `Dockerfile`- any tool which can build container images should work.

**Comment by cgwalters** [comment]
The initial version of this is implemented in the [0.4.0 release](https://docs.rs/ostree-ext/0.4/ostree_ext/container/store/index.html].  Closing this issue since we have a working PoC - a lot more remains to be done obviously, but will make a followup tracker issue.
